File: src/main/java/chat/teco/tecochat/comment/application/CommentService.java
package chat.teco.tecochat.comment.application;

import chat.teco.tecochat.chat.domain.chat.Chat;
import chat.teco.tecochat.chat.domain.chat.ChatRepository;
import chat.teco.tecochat.comment.application.dto.DeleteCommentCommand;
import chat.teco.tecochat.comment.application.dto.UpdateCommentCommand;
import chat.teco.tecochat.comment.application.dto.WriteCommentCommand;
import chat.teco.tecochat.comment.domain.Comment;
import chat.teco.tecochat.comment.domain.CommentRepository;
import chat.teco.tecochat.domain.member.MemberRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Transactional
@Service
public class CommentService {

    private final MemberRepository memberRepository;
    private final ChatRepository chatRepository;
    private final CommentRepository commentRepository;

    public Long write(WriteCommentCommand command) {
        memberRepository.getById(command.memberId());
        Comment comment = command.toDomain();
        Chat chat = chatRepository.getById(command.chatId());
        chat.increaseComment();
        return commentRepository.save(comment).id();
    }

    public void update(UpdateCommentCommand command) {
        Comment comment = commentRepository.getById(command.commentId());
        comment.update(command.memberId(), command.content());
    }

    public void delete(DeleteCommentCommand command) {
        Comment comment = commentRepository.getById(command.commentId());
        comment.validateDelete(command.memberId());
        Chat chat = chatRepository.getById(comment.chatId());
        chat.decreaseComment();
        commentRepository.delete(comment);
    }
}

File: src/main/kotlin/chat/teco/tecochat/application/CommentService.kt
package chat.teco.tecochat.application

import chat.teco.tecochat.domain.chat.ChatRepository
import chat.teco.tecochat.domain.comment.Comment
import chat.teco.tecochat.domain.comment.CommentRepository
import chat.teco.tecochat.domain.member.MemberRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Transactional
@Service
class CommentService(
    private val chatRepository: ChatRepository,
    private val commentRepository: CommentRepository,
    private val memberRepository: MemberRepository,
) {

    fun write(memberId: Long, request: WriteCommentRequest): Long {
        val comment = request.toComment(memberId)
        val chat = chatRepository.getById(request.chatId)
        chat.increaseComment()
        return commentRepository.save(comment).id
    }

    fun update(memberId: Long, commentId: Long, updateCommentRequest: UpdateCommentRequest) {
        val comment = commentRepository.getById(commentId)
        comment.update(memberId, updateCommentRequest.content)
    }

    fun delete(memberId: Long, commentId: Long) {
        val comment = commentRepository.getById(commentId)
        comment.validateDelete(memberId)
        val chat = chatRepository.getById(comment.chatId)
        chat.decreaseComment()
        commentRepository.delete(comment)
    }

    fun findAllByChatId(chatId: Long): List<CommentResponse> {
        val comments = commentRepository.findAllByChatId(chatId)
        val members = memberRepository.findAllById(comments.map(Comment::memberId))
            .associateBy { it.id }
        return comments.map { CommentResponse(it, members[it.memberId]) }
    }
}

